include/swift/AST/DiagnosticConsumer.h

```diff
@@ -111,6 +112,35 @@ class NullDiagnosticConsumer : public DiagnosticConsumer {
                         ArrayRef<DiagnosticArgument> FormatArgs,
                         const DiagnosticInfo &Info) override;
 };
+
+/// \brief DiagnosticConsumer that funnels diagnostics in certain files to
+/// particular sub-consumers.
+///
+/// Diagnostics that are not in one of the special files are emitted into every
+/// sub-consumer.
+class FileSpecificDiagnosticConsumer : public DiagnosticConsumer {
+public:
+  using ConsumerPair =
+      std::pair<CharSourceRange, std::unique_ptr<DiagnosticConsumer>>;
+
+private:
+  SmallVector<std::unique_ptr<DiagnosticConsumer>, 4> SubConsumers;
+  SmallVector<std::pair<CharSourceRange, unsigned>, 4> LocationToConsumerMap;
+  static const unsigned NO_SPECIFIC_CONSUMER = -1;
+  unsigned CurrentGroupConsumer = NO_SPECIFIC_CONSUMER;
```

Comment:
Would this work as a local variable inside of `handleDiagnostic`? I don't see other uses of it, and AFAICT it is set each time that routine is called.

Reply:
Ah--now I see it is set for non-notes and reused for notes. A better name would be `GroupConsumerForPreviousNonNote`. No, that's not quite right. At least a Doxygen comment to explain it would help.

Reply:
Glad you like the idea. Tx.